Add benchmarks to evaluate overhead

brawner opened the pull request:
Based on some feedback from #10 I wanted to measure overhead of different choices with this infrastructure. This adds malloc/realloc benchmarks to measure the overhead of tracking them. It also enables linters for this package. 

Signed-off-by: Stephen Brawner <brawner@gmail.com>

brawner committed:
Add benchmarks to evaluate overhead

Signed-off-by: Stephen Brawner <brawner@gmail.com>

brawner commented:
Building with benchmarks, `--packages-select performance_test_fixture`

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=13025)](http://ci.ros2.org/job/ci_linux/13025/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7975)](http://ci.ros2.org/job/ci_linux-aarch64/7975/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=10751)](http://ci.ros2.org/job/ci_osx/10751/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=13007)](http://ci.ros2.org/job/ci_windows/13007/)

cottsay commented on CMakeLists.txt:
```diff
@@ -48,7 +49,21 @@ install(
+  find_package(ament_lint_auto REQUIRED)
+  ament_lint_auto_find_test_dependencies()
+
+  set(_PERFORMANCE_TEST_FIXTURE_TESTING_IN_PROJECT TRUE)
```
You can drop all of the special casing you added to `add_performance_test.cmake` and `performance_test_fixture-extras.cmake` if you do this instead:
```suggestion
  set(performance_test_fixture_DIR "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
  add_library(${PROJECT_NAME}::${PROJECT_NAME} ALIAS ${PROJECT_NAME})
```

brawner replied:
I thought about just manually setting the `_DIR` variable, but the add_library bit was what I was missing. Awesome.

cottsay commented on test/benchmark/benchmark_malloc_realloc.cpp:
```diff
@@ -0,0 +1,144 @@
+BENCHMARK_REGISTER_F(PerformanceTestFixture, benchmark_on_realloc)
+->ArgNames({"Enable Performance Tracking", "Malloc Size", "Realloc Size"})
+->Apply(realloc_args);
+
+BENCHMARK_MAIN();
```
We don't explicitly do this in any other tests. Is there a particular reason it's done here?

```suggestion
```

brawner replied:
Removed

cottsay replied:
Doesn't appear to have been removed yet.

brawner replied:
Actually removed

cottsay commented on test/benchmark/benchmark_malloc_realloc.cpp:
```diff
@@ -0,0 +1,144 @@
+BENCHMARK_REGISTER_F(PerformanceTestFixture, benchmark_on_malloc)
+->ArgNames({"Enable Performance Tracking", "Malloc Size"})->Apply(malloc_args);
+
+// Three types of realloc tests, one where malloc is smaller than realloc, one where they are
+// the same, and one where malloc is larger than realloc. Realloc size ranges from 1 to 2^27
+// each time multiplying by 32. Each stop is tested with/without performance metrics
+static void realloc_args(benchmark::internal::Benchmark * b)
+{
+  for (int malloc_adjustment = -1; malloc_adjustment <= 1; ++malloc_adjustment) {
+    for (int realloc_shift = 0; realloc_shift < 32; realloc_shift += 8) {
+      for (int enable_tracking = kDisablePerformanceTracking;
+        enable_tracking <= kEnablePerformanceTracking; ++enable_tracking)
+      {
+        const int malloc_shift = realloc_shift + malloc_adjustment;
+        if (malloc_shift < 0) {
+          continue;
+        }
+        b->Args({enable_tracking, 1 << malloc_shift, 1 << realloc_shift});
+      }
+    }
+  }
+}
```
Just curious - have you tried using the `->Ranges({ ... })` function to do this instead?

brawner replied:
I did look at that, but I couldn't find a way to control the spacing when using `Ranges`. Instead benchmark chooses values itself? If I use ranges instead, it generates 22 benchmarks when using a similar range, but they are reasonable spaced. While the number is a bit high, I was most worried about not having control over the benchmark it actually does generate. It seems like we might be at its mercy if we update the vendored version or something?

brawner replied:
I cleaned up both functions by removing the inner most for loop. It was added originally on the outside, but it can be unrolled on the inside.

cottsay replied:
Makes sense, thanks!

brawner replied:
Also, it looks like if you have `Ranges` with two lists, it will run all combinations. In the new benchmark for testing without memory tools, I tried `{{1, 1<<27}, {2, 1 << 28}}` and it chose to run so many that the benchmark ultimately timed out.

cottsay reviewed (changes requested):
A couple of comments. Also, please re-run CI with `AMENT_RUN_PERFORMANCE_TESTS=1` to see what the results are like.

brawner committed:
PR fixup

Signed-off-by: Stephen Brawner <brawner@gmail.com>

brawner commented:
Addressed the msbuild warnings, and running this time with the benchmarks actually on (hopefully).

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=13029)](http://ci.ros2.org/job/ci_linux/13029/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7979)](http://ci.ros2.org/job/ci_linux-aarch64/7979/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=10755)](http://ci.ros2.org/job/ci_osx/10755/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=13011)](http://ci.ros2.org/job/ci_windows/13011/)

brawner commented:
Benchmark results can be seen here:
https://ci.ros2.org/job/ci_linux/13029/testReport/(root)/projectroot/benchmark_malloc_realloc/

cottsay reviewed (approved):
Looks good once `BENCHMARK_MAIN` has been dropped.

cottsay commented:
This is great. I'd like to note, however, that the `LD_PRELOAD` is still in effect even though the memory tools aren't active, so there is still additional overhead associated with memory operations. I'm showing 270/70/12 ns for testing with memory tools, with memory tools disabled, and without preloading (respectively).

You'd have to use a separate test and add it with `ament_add_google_benchmark`, but you could test that here if you'd like. It's up to you.

brawner committed:
Add malloc/realloc test without memory tools

Signed-off-by: Stephen Brawner <brawner@gmail.com>

brawner commented:
Added a new test for use without memory tools. I also added logic so that it would be skipped if the other benchmark is skipped.

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=13040)](http://ci.ros2.org/job/ci_linux/13040/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7988)](http://ci.ros2.org/job/ci_linux-aarch64/7988/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=10764)](http://ci.ros2.org/job/ci_osx/10764/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=13020)](http://ci.ros2.org/job/ci_windows/13020/)

brawner committed:
Fix int conversions

Signed-off-by: Stephen Brawner <brawner@gmail.com>

brawner commented:
I think I fixed the warnings above
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=13041)](http://ci.ros2.org/job/ci_linux/13041/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7989)](http://ci.ros2.org/job/ci_linux-aarch64/7989/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=10765)](http://ci.ros2.org/job/ci_osx/10765/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=13021)](http://ci.ros2.org/job/ci_windows/13021/)

brawner commented:
Without memory tools. I'm impressed by the memory latency of this build on cloud VMs.
```
---------------------------------------------------------------------------------------------------------------------------
Benchmark                                                                                 Time             CPU   Iterations
---------------------------------------------------------------------------------------------------------------------------
BenchmarkFixture/benchmark_on_malloc/Malloc Size:1                                     9.53 ns         9.53 ns     73326267
BenchmarkFixture/benchmark_on_malloc/Malloc Size:16                                    10.2 ns         10.2 ns     69127812
BenchmarkFixture/benchmark_on_malloc/Malloc Size:256                                   10.2 ns         10.2 ns     69155989
BenchmarkFixture/benchmark_on_malloc/Malloc Size:4096                                  23.2 ns         23.2 ns     30191050
BenchmarkFixture/benchmark_on_malloc/Malloc Size:65536                                 27.4 ns         27.4 ns     25594158
BenchmarkFixture/benchmark_on_malloc/Malloc Size:1048576                               25.3 ns         25.3 ns     27656313
BenchmarkFixture/benchmark_on_malloc/Malloc Size:16777216                              25.3 ns         25.3 ns     27711868
BenchmarkFixture/benchmark_on_malloc/Malloc Size:268435456                             4804 ns         4804 ns       145695
```

brawner committed:
Int literals

Signed-off-by: Stephen Brawner <brawner@gmail.com>

brawner commented:
Windows is picky about the shift conversion, rebuilding after latest commit
* [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_windows&build=13022)](https://ci.ros2.org/job/ci_windows/13022/)

cottsay commented on CMakeLists.txt:
```diff
@@ -48,7 +49,43 @@ install(
+    test/benchmark/benchmark_malloc_realloc.cpp
+    TIMEOUT 60)
+
+  # Adding executable and linking done manually to avoid dependency on
+  # osrf_testing_tools_cpp
+  if(TARGET benchmark_malloc_realloc)
+    ament_add_google_benchmark_executable(
+      benchmark_malloc_realloc_no_memory_tools
+      test/benchmark/benchmark_malloc_realloc_no_memory_tools.cpp)
+    if(TARGET benchmark_malloc_realloc_no_memory_tools)
+      target_link_libraries(
+        benchmark_malloc_realloc_no_memory_tools
+        benchmark::benchmark)
+
+      get_target_property(ARG_SKIP_RETURN_CODE benchmark_malloc_realloc SKIP_RETURN_CODE)
+      if(ARG_SKIP_RETURN_CODE)
+        set(ARG_SKIP_TEST "SKIP_TEST")
+      endif()
+
+      ament_add_google_benchmark_test(
+          benchmark_malloc_realloc_no_memory_tools ${ARG_SKIP_TEST})
+    endif()
+  endif()
```
Since `ament_add_google_benchmark` is just a vanilla Google Benchmark test, it has nothing to do with `osrf_testing_tools_cpp`. So you can use the existing macros, and this whole block gets a lot more simple:
```suggestion
  # Bypassing performance_test_fixture to avoid LD_PRELOAD of
  # osrf_testing_tools_cpp
  set(_no_mem_tools_skip_arg)
  if(NOT TARGET benchmark_malloc_realloc)
    set(_no_mem_tools_skip_arg SKIP_TEST)
  endif()
  ament_add_google_benchmark(
    benchmark_malloc_realloc_no_memory_tools
    test/benchmark/benchmark_malloc_realloc_no_memory_tools.cpp
    ${_no_mem_tools_skip_arg})
```

brawner replied:
I like the simpler logic, but I also added a clause to check if benchmark_malloc_realloc is being skipped, then this one should be too.

cottsay commented on test/benchmark/benchmark_malloc_realloc_no_memory_tools.cpp:
```diff
@@ -0,0 +1,91 @@
+
+#include "./macros.h"
+
+using BenchmarkFixture = ::benchmark::Fixture;
```
This seems like a roundabout way to essentially use no fixture at all. The standard way looks like this: https://github.com/google/benchmark#basic-usage

Unless there is a specific reason you're doing it like this?

brawner replied:
I guess just too much copy and paste. Adjusted.

cottsay reviewed (changes requested):
A couple of thoughts on the new stuff.

brawner committed:
PR feedback

Signed-off-by: Stephen Brawner <brawner@gmail.com>

cottsay reviewed (approved):
Should probably get one last round of CI, but this looks good to me. Thanks for iterating.

brawner commented:
Hopefully the last round of CI

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=13042)](http://ci.ros2.org/job/ci_linux/13042/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7990)](http://ci.ros2.org/job/ci_linux-aarch64/7990/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=10766)](http://ci.ros2.org/job/ci_osx/10766/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=13023)](http://ci.ros2.org/job/ci_windows/13023/)

brawner commented:
Warnings on aarch and windows are just the expected ones for benchmarks.
